README.md
AtlasMapper
===========

The AtlasMapper is a cross platform JavaScript/Java mapping application that allows a catalogue of map layers (WMS, NCWMS, ArcGIS Server, XYZ / OSM tiles, Google, Bing, basic KML and eventually other formats supported by OpenLayers) to be easily browsed, layered, re-styled and compared side-by-side in a web browser. It consists of a Java servlet server for layer catalogue management and a Javascript client based on GeoExt, Ext-JS and OpenLayers.

Demo
----

To see what the mapping client looks like try out the demo version of the client. The AtlasMapper server allows this client to be configured with Web Map Server data sources. Try out the demo version of the server (username: admin, password: admin).

Note: The demo can only connect to web services that are on the public internet and so will not work with those hosted on localhost. To work with localhost services download the AtlasMapper and run it locally.

More info
---------

For more info, consult the <a href="http://atlasmapper.org/">AtlasMapper website</a>.

Building the AtlasMapper from Source
------------------------------------

This section will eventually describe the building of the source to produce an executable WAR file.


Changelog.md
# 2015-07-02: Version 1.5.5 #
> Fixed issues
    * Fix bug introduced in 1.5.4 (URL parameters double encoded)
    * Fix "Legend Parameters" field, always apply those parameters to the legend even when they are defined with the layer tiles.
    * Print frame fix "Coordinate lines" scroller label (it used to be "Scale lines")
> New feature
    * Print frame: Keep widgets near the print frame (the North arrow and the scale line)

# 2015-06-25: Version 1.5.4 #
> Client
    * Added 2 sliders to the print frame option panel, to configure the coordinate lines around the print frame.

# 2015-06-23: Version 1.5.3 #
> Fixed issues
    * Fix [issue #49](https://code.google.com/p/atlasmapper/issues/detail?id=#49) : Feature request do not work when browser zoom is not 100%. The bug was resolved in ver 1.4 and somehow it came back...
    * Fix legend duplicated parameters (which only cause problem when TRANSPARENT=FALSE is set for legend parameters in the data source configuration)
    * Fix documentation explaining how to set ncWMS default colour scale, and colour scale field label (typo)
    * Fix endless loading event with print frame and location search

# 2014-09-04: Version 1.5.1 #
> Fixed issues
    * [Issue 99](https://code.google.com/p/atlasmapper/issues/detail?id=99): Features Requests not working with OpenLayers 2.13

# 2014-08-20: Version 1.5.0 #
> Fixed issues
    * Fixed touch device "double" zoom (removed PinchZoom control; it's already included in Navigation)
    * Fixed IE bug with parameter parsing (IE "split" method do not follow ECMA Script specification, it ignores empty value)
    * Fixed IE 10 tile scramble bug (tiles get completely random on IE 10 only, after a certain number of zooming events - fixed by upgrading to OpenLayers 2.13.1)
    * Fixed a few Java warnings and potentials Null pointer exceptions.

# 2013-12-24: Version 1.4.8 #
> Fixed issues
    * [Issue 26](https://code.google.com/p/atlasmapper/issues/detail?id=26): Feature selection popup ballons show unnecessary scrollbars in Chrome
    * [Issue 28](https://code.google.com/p/atlasmapper/issues/detail?id=28): Feature selection only works on top layer - Often not appropriate
    * [Issue 55](https://code.google.com/p/atlasmapper/issues/detail?id=55): High DPI mode doesn't work with gutter oloptions
    * [Issue 65](https://code.google.com/p/atlasmapper/issues/detail?id=65): Layer re-order sometime fail when moving to the top
    * [Issue 67](https://code.google.com/p/atlasmapper/issues/detail?id=67): HTML / JavaScript injection (Security)
    * [Issue 68](https://code.google.com/p/atlasmapper/issues/detail?id=68): NullPointerException when generating a client without default layers
    * [Issue 70](https://code.google.com/p/atlasmapper/issues/detail?id=70): Misleading error reporting for WMS source with no WMTS service.
    * [Issue 71](https://code.google.com/p/atlasmapper/issues/detail?id=71): Null pointer exception while generating client with none generated data source
    * [Issue 73](https://code.google.com/p/atlasmapper/issues/detail?id=73): Square brackets in download links appear encoded to user.
    * [Issue 75](https://code.google.com/p/atlasmapper/issues/detail?id=75): Typo in Documentation
    * [Issue 77](https://code.google.com/p/atlasmapper/issues/detail?id=77): "Enable Web cache" do not work as expected
    * [Issue 79](https://code.google.com/p/atlasmapper/issues/detail?id=79): Removing a layer while other layers are loading removes loading status
> Server
    * Added configuration for measurement tools and download service
    * TC211 Parser: Better distinction between normal resources and downloadable resources
      * Download resources are only shown in the download window
> Client
    * Added area measurement tool and fixed measurement line style
    * Extended wiki-format support (the wiki format will soon be move server side and the logic will change)
    * New feature requests system (see [issue #28](https://code.google.com/p/atlasmapper/issues/detail?id=#28))
    * Download window
      * Added a download logger service capability (configurable with the admin)
      * Added a close button
    * Doctype changed to HTML5
    * Added a crash warning for IE 6

# 2013-11-4: Version 1.4.7 #
> Fixed issues
    * [Issue 65](https://code.google.com/p/atlasmapper/issues/detail?id=65): Layer re-order sometime fail when moving to the top
    * [Issue 55](https://code.google.com/p/atlasmapper/issues/detail?id=55): High DPI mode doesn't work with gutter oloptions
    * [Issue 26](https://code.google.com/p/atlasmapper/issues/detail?id=26): Feature selection popup ballons show unnecessary scrollbars in Chrome
> Server
    * Improve transfer speed
      * Added ZIP capabilities for files requests going through the AtlasMapper
      * Fixed response header to tell the browsers how to cache the files
    * Fixed issue related to errors not reported correctly when generating a client
    * Implemented basic layer search (without indexation)
    * Client configuration - Data sources now displayed in a grid, on a separate tab
> Client
    * Added layer search "beta" (using basic regex server side, without indexation)
    * Added measurement tool "beta" (not configurable - in the future, the tool will create a configurable "layer")
    * Hide legend (WMS) that contains no images (or broken images)
    * Implement Hi-DPI with KML
    * Fixed usability issue related to location search results on the map (the icon used to highlight the result on mouseover. This functionality can not be used since it imply that all vector layers would have mouse over for feature requests, including KML. The click event on search results was not implemented correctly)
    * Disabled mouse wheel on embedded map
    * Fixed some issues with the Wiki format (the style delimiter regex has been redesigned to support more cases)
    * Fixed the look of the coordinate box (mouse position) on the bottom right corner of the map

src/main/webapp/javascript/Frameset.js
/*
 *  This file is part of AtlasMapper server and clients.
 *
 *  Copyright (C) 2011 Australian Institute of Marine Science
 *
 *  Contact: Gael Lafond <g.lafond@aims.org.au>
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
// Add the additional 'advanced' validation type (VTypes)
// to validate json syntaxe.
// ExtJS implement a nice tool to decode JSON. Unfortunatly it's errors
// description are useless.

Ext.define('Frameset', {
	extend: 'Ext.container.Viewport',

	layout: 'border',
	contentPanel: null,

	waitingTitle: null,
	waitingContentItems: null,

	defaults: {
		split: true
	},

	statusBar: null,
	statusWindow: null,

	initComponent: function() {
		var contentItems = [];
		if (this.waitingContentItems) {
			contentItems = this.waitingContentItems;
			this.waitingContentItems = null;
		}

		var that = this;
		var contentPanelConfig = {
			autoScroll: true,
			collapsible: false,
			region:'center',
			bodyPadding: 10,
			items: contentItems,

			// - Items takes 100% of horizontal space by default -
			//   Without this, items takes 100% after render,
			//   but do not adjust their width on window resize.
			layout: 'anchor'
		};

		if (this.waitingTitle) {
			contentPanelConfig.title = this.waitingTitle;
			this.waitingTitle = null;
		}

		this.contentPanel = Ext.create('Ext.panel.Panel', contentPanelConfig);

		var logoutHTML = '<p style="float: right">';
		if (userName) {
			logoutHTML += 'Logged as <a href="../admin/userAccountPage.jsp">' + userName + '</a>. ';
		}
		logoutHTML += '<a href="../public/login.jsp?action=logout">[Logout]</a></p>';

		this.statusBar = Ext.create('Ext.ux.StatusBar', {
			height: 25,
			minHeight: 25,
			maxHeight: 250,
			region: 'south',
			defaultText: 'Ready'
		});

		this.items = [
			{
				html: logoutHTML +
					'<p><img src="../resources/images/AtlasMapper_logo_shadow_252x50px.png" /></p>',
				border: false,
				region: 'north',
				height: 70,
				minHeight: 50,
				maxHeight: 250,
				bodyPadding: '5',
				cmargins: '0 0 5 0'
			}, {
				html: '<ul class="bullet-list">'+
						'<li><a href="../admin/dataSourcesConfigPage.jsp">Data sources</a></li>'+
						'<li><a href="../admin/clientsConfigPage.jsp">AtlasMapper clients</a></li>'+
						'<li><a href="../admin/manualOverrideDoc.html" target="_blank">Documentation</a></li>'+
						'<li><a href="../admin/aboutPage.jsp">About</a></li>'+
					'</ul>',
				title: 'Navigation',
				region:'west',
				width: 200,
				minWidth: 100,
				maxWidth: 250,
				cmargins: '0 5 0 0'
			},
			this.contentPanel,
			this.statusBar
		];

		this.callParent(arguments);
	},

	setContentTitle: function(contentTitle) {
		if (!this.contentPanel) {
			this.waitingTitle = contentTitle;
		} else {
			this.contentPanel.setTitle(contentTitle);
		}
	},

	addContentDescription: function(contentDesc) {
		this.addContent({
			xtype: 'panel',
			margin: '0 0 15 0',
			border: false,
			html: contentDesc
		});
	},

	addContent: function(contentItem) {
		if (!this.contentPanel) {
			// Add the content to the waiting list if the panel is not ready.
			if (!this.waitingContentItems) {
				this.waitingContentItems = [];
			}
			this.waitingContentItems.push(contentItem);
		} else {
			this.contentPanel.add(contentItem);
		}
	},


	redirectIfNeeded: function(statusCode) {
		// 401 = UNAUTHORIZED
		if (statusCode == 401) {
			Ext.Msg.show({
				 title: 'Failure',
				 msg: 'Your session has timed out.<br/>Please wait...',
				 icon: Ext.Msg.WARNING
			});
			// Redirect to the login page
			window.location = '../public/admin.jsp';
			return true;
		}
		return false;
	},

	beforeShow: function() {
		if (this.statusWindow) {
			this.statusWindow.close();
			this.statusWindow = null;
		}
	},

	showBusy: function() {
		this.beforeShow();
		this.statusBar.showBusy();
	},

	/**
	 * Display the error message in the status bar and popup a window
	 * using the error message as a title and the errors from the
	 * response as content, formated using a bullet list.
	 * msg: A brief error message
	 * response: The response object contains errors
	 */
	setErrors: function(msg, response, statusCode) {
		this.beforeShow();
		if (!this.redirectIfNeeded(statusCode)) {
			var errorMessages = '';
			if (response && response.errors) {
				errorMessages = this.errorsToHtml(response.errors);
			}

			var alertMsg = '<b>' + msg + '</b>';
			if (errorMessages) {
				alertMsg += '<br/>\n' + errorMessages;
			}
			Ext.Msg.alert('Failure', alertMsg);
			this.statusBar.setStatus({
				text: msg,
				iconCls: 'x-status-error',
				clear: true // auto-clear after a set interval
			});
		}
	},
	setErrorsAndWarnings: function(title, msg, response) {
		this.beforeShow();
		var errorMessages = '';

		if (response && response.messages) {
			errorMessages += '<div><b>Information:</b>';
			errorMessages += this.errorsToHtml(response.messages);
			errorMessages += '</div>\n';
		}

		if (response && response.errors) {
			errorMessages += '<div><b>Errors:</b>';
			errorMessages += this.errorsToHtml(response.errors);
			errorMessages += '</div>\n';
		}

		if (response && response.warnings) {
			errorMessages += '<div><b>Warnings:</b>';
			errorMessages += this.errorsToHtml(response.warnings);
			errorMessages += '</div>\n';
		}

		var alertMsg = '';
		if (msg) {
			alertMsg += '<b>' + msg + '</b><br/>\n<br/>\n';
		}
		if (errorMessages) {
			alertMsg += errorMessages;
		}

		Ext.create('Ext.window.Window', {
			layout: 'fit',
			title: title,
			closable: true,
			resizable: true,
			plain: true,
			border: false,
			width: 550,
			height: 400,
			items: [{
				xtype: 'panel',
				autoScroll: true,
				bodyPadding: 5,
				html: alertMsg
			}],
			dockedItems: [{
				xtype: 'toolbar',
				dock: 'bottom',
				ui: 'footer',
				layout:{
					pack: 'center'
				},
				defaults: { minWidth: 75 },
				items: [
					{
						xtype: 'button',
						text: 'Close',
						padding: '2 10',
						handler: function() {
							var window = this.ownerCt.ownerCt;
							window.close();
						}
					}
				]
			}]
		}).show();

		this.statusBar.setStatus({
			text: msg,
			iconCls: 'x-status-error',
			clear: true // auto-clear after a set interval
		});
	},
	errorsToHtml: function(errors) {
		var errorMessages = '<ul class="bullet-list">\n';
		if (Ext.isArray(errors)) {
			Ext.each(errors, function(message) {
				if (typeof(message) === 'string') {
					errorMessages += '<li>' + this.errorMessageToHtml(message) + '</li>\n';
				} else {
					Ext.iterate(message, function(url, messageStr) {
						errorMessages += '<li>';
						if (typeof(messageStr) === 'string') {
							errorMessages += this.errorMessageToHtml(messageStr);
						} else {
							var first = true;
							Ext.each(messageStr, function(messageElStr) {
								if (first) { first = false; } else { errorMessages += '<br/>\n';}
								errorMessages += this.errorMessageToHtml(messageElStr);
							}, this);
						}
						if (url) {
							errorMessages += '<br/>\n' + this.urlToHtmlLink(url);
						}
						errorMessages += '</li>\n';
					}, this);
				}
			}, this);
		} else {
			Ext.iterate(errors, function(clientName, messageArray) {
				errorMessages += '<li>' + clientName;
				errorMessages += this.errorsToHtml(messageArray);
				errorMessages += '</li>';
			}, this);
		}
		errorMessages += '</ul>\n';

		return errorMessages;
	},
	errorMessageToHtml: function(str) {
		return Ext.String.htmlEncode(str).replace(/\n/g, "<br/>\n");
	},
	urlToHtmlLink: function(url) {
		var urlDisplay = url;

		var maxUrlLength = 60;
		if (maxUrlLength > 0 && maxUrlLength < url.length) {
			var beginningLength = Math.round((maxUrlLength-3) * 3.0/4);
			var endingLength = maxUrlLength - beginningLength - 3; // 3 is for the "..."
			if (beginningLength > 1 && endingLength == 0) {
				beginningLength--;
				endingLength = 1;
			}
			urlDisplay = url.substring(0, beginningLength) + "..." + url.substring(url.length - endingLength);
		}

		return '<a href="'+url+'" target="_blank">'+Ext.String.htmlEncode(urlDisplay)+'</a>';
	},

	setError: function(error, statusCode) {
		this.beforeShow();
		if (!this.redirectIfNeeded(statusCode)) {
			Ext.Msg.alert('Failure', error);
			this.statusBar.setStatus({
				text: error,
				iconCls: 'x-status-error',
				clear: true // auto-clear after a set interval
			});
		}
	},

	setValidMessage: function(status) {
		this.beforeShow();
		this.statusBar.setStatus({
			text: status,
			iconCls: 'x-status-valid',
			clear: true // auto-clear after a set interval
		});
	},

	setSavingMessage: function(msg) {
		this.beforeShow();
		this.statusWindow = Ext.Msg.show({
			title: msg,
			msg: 'Please wait...',
			closable: false,
			icon: Ext.Msg.INFO,
			closeAction: 'destroy'
		});

		this.statusBar.setStatus({
			text: msg,
			iconCls: 'x-status-saving'
		});
	},

	setSavedMessage: function(msg, delay, windowTitle) {
		this.beforeShow();

		if (typeof(windowTitle) != 'undefined') {
			this.statusWindow = Ext.Msg.show({
				title: windowTitle,
				msg: msg,
				icon: Ext.Msg.INFO,
				closeAction: 'destroy'
			});
		}

		var config = {
			text: msg,
			iconCls: 'x-status-saved',
			clear: true // auto-clear after a set interval
		};
		if (delay && delay > 0) {
			var that = this;
			Ext.defer(function() {
				that.statusBar.setStatus(config);
			}, delay);
		} else {
			this.statusBar.setStatus(config);
		}
	},

	setTextMessage: function(status) {
		this.beforeShow();
		this.statusBar.setStatus({
			text: status,
			iconCls: 'x-status-text',
			clear: true // auto-clear after a set interval
		});
	},

	clearStatus: function() {
		this.beforeShow();
		this.statusBar.clearStatus({useDefaults:true});
	}
});
